### PR TITLE
Import Activity Reports from smartsheets

### PR DESCRIPTION
**Pull request by dcloud**
**Description of change**

This adds a script to import activity reports from CSV files. If you have a collection of `R\d{1,2}*.csv` CSV files (in the format exported from smartsheets), you should be able to upload those to s3 (minio) and `yarn import:reports:local --file file.csv --region 14` and import CSV data. Check out `src/tools/importActivityReports.js` for the bulk of the work.

There is a JSONB field, `imported`, for storing information from the parsed CSV, which would allow for future improvements to imported data. The import process attempts to convert rows into records, with various caveats:

- The importer is set to skip rows that don't have a report id, since those are likely blank rows.
- The import does not many address relational fields. For example, User records won't be created prior to import, and there is no guarantee that the emails from smartsheets will match up.
- Grant/Grantee records. If a Grant with a matching name can be found, an `ActivityRecipient` will be created.
- `{ validate: false }`, because `checkRequiredForSubmission` checks for values that may not exist during import
- Records that started with *R14* were given a `legacyId` that uses the actual region id (taken from the filename) rather than the test region 14. The `regionId` is taken from `--region` cli option and used to correct the region in the identifier. You can still get the old value via `"imported"->'reportId'`.
- Depending on how you convert from Excel to csv (save as from Excel, use `csvkit`), date fields in the output may differ. We try to address that in `coerceDate`, but YMMV.

**How to test**

- Get your hands on some Smartsheet files, convert to csv, upload them to minio.
- `yarn db:migrate` to remove some "not null" requirements, create the "imported" JSONB column
- run `yarn import:reports:local --file R1ActivityReports.csv --region 1` or similar.
- Check the db: `SELECT count(id) FROM "ActivityReports" WHERE "legacyId" IS NOT NULL`, for example

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/215

**Checklist**

- [ ] Meets issue criteria
- [ ] Code tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

File: src/services/activityReports.js [hunk 1]
```diff
@@ -171,7 +171,7 @@ export async function review(report, status, managerNotes) {
 
 export function activityReportById(activityReportId) {
   return ActivityReport.findOne({
-    attributes: { exclude: ['legacyId'] },
+    attributes: { exclude: ['imported', 'legacyId'] },
```
Comment (dcloud) [inline src/services/activityReports.js:174]: I may need to undo that if we're going to make a frontend view of the un-processed imported data.

**Comment by rahearn** [comment]
Talking about this with @dcloud. We decided on the following:

Things we need right away:

* a legacy view-only mode that displays data from of the import json
* Matching grant numbers in the imported data to existing grants. If none found, continue to leave blank

Things we need to do soon, but post-mvp

* create a job to match user emails (owners, collaborators, and approving mangers) in the import json to newly created user emails in the User table
* Ryan to import comments and attachments from smartsheet into the hub

As users start to login in the first week, old ARs will be missing attachments, and user relationships will be filled in as users register.
There may be other mismatches between valid form values in Smartsheet vs Smart Hub, we will reconcile those as best as possible through later enhancements.

**Comment by dcloud** [comment]
Putting this in "ready for review" status. The import script is missing a test.js file, but I'll need some time to generate a testable CSV. Otherwise, I think the tires are ready for kicking, and thanks to @jasalisbury there is a custom legacy report view/route so you can look at things on the frontend.

**Review by rahearn — CHANGES_REQUESTED** [review]

File: frontend/src/fetchers/activityReports.js [hunk 2]
```diff
@@ -5,6 +5,11 @@ import { DECIMAL_BASE } from '../Constants';
 const activityReportUrl = join('/', 'api', 'activity-reports');
 const activityReportAlertUrl = join('/', 'api', 'activity-reports', 'alerts');
 
+export const legacyReportById = async (legacyId) => {
+  const res = await get(join(activityReportUrl, 'legacy', legacyId.toString(DECIMAL_BASE)));
```
[the remark this answers is not in the document]
Comment (jasalisbury) [reply frontend/src/fetchers/activityReports.js:9]: Yeah good catch

**Comment by dcloud** [comment]
I'll note that this is essentially done. The test I added for the import command is fairly minimal, and could be expanded on. However, tests in other files were failing for me locally, so I'd like to have CircleCI (and other reviewers, possibly) verify that the new tests aren't breaking others (since they do create AR database records).

File: .gitignore [hunk 3]
```diff
@@ -25,3 +25,6 @@ node_modules
 .DS_Store
 .env
 **/*secrets.auto.tfvars
+
+# Temp files
+.tmp
```
[the remark this answers is not in the document]
Comment (dcloud) [reply .gitignore:30]: Addressed via 0d76df0

File: src/tools/importActivityReports.test.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    expect(records).toBeDefined();
+    expect(records.length).toBe(10);
+
+    expect(records).toContainEqual(
+      expect.objectContaining({ id: expect.anything(), legacyId: 'R14-AR-000279', requester: 'Regional Office' }),
+    );
+
+    expect(records).toContainEqual(
+      expect.objectContaining({ id: expect.anything(), legacyId: 'R14-AR-001132', requester: 'Grantee' }),
```
Comment (rahearn) [inline src/tools/importActivityReports.test.js:43]: Ooh, I like these matchers. That's great.

**Review by rahearn — APPROVED** [review]
Really impressed with the import and display code here. Great work

**Comment by kryswisnaskas** [comment]
Looks good! Verified I was able to import all regions and view the individual reports under the legacy endpoint.

The only issue I saw was the display of "Status" in some of the reports. Suggested change to the Landing/index.css

```
.smart-hub--status- {
  display: none;
}

.smart-hub--status-null {
  display: none;
}
```
While this might not be the most elegant solution, it appears to be the simplest.

**Comment by jasalisbury** [comment]
> Looks good! Verified I was able to import all regions and view the individual reports under the legacy endpoint.
> 
> The only issue I saw was the display of "Status" in some of the reports. Suggested change to the Landing/index.css
> 
> ```
> .smart-hub--status- {
>   display: none;
> }
> 
> .smart-hub--status-null {
>   display: none;
> }
> ```
> 
> While this might not be the most elegant solution, it appears to be the simplest.

After looking a bit at this I realized we only show reports that are approved in the main reports table. This means imported reports that are not approved will not show up in either reports table. A user will have to know the legacy ID to view a legacy report that has not been approved.

**Comment by rahearn** [comment]
> A user will have to know the legacy ID to view a legacy report that has not been approved.

Yes, for now.

> Things we need to do soon, but post-mvp
> create a job to match user emails (owners, collaborators, and approving mangers) in the import json to newly created user emails in the User table

Once this is done non-approved reports will start appearing in the appropriate user alert tables

**Review by kryswisnaskas — APPROVED** [review]

**Pull request merged**